proposals/ref-readonly-parameters.md

```diff
@@ -0,0 +1,176 @@
+`in` parameters allow both lvalues and rvalues and can be used without any annotation at the callsite.
+However, APIs which capture or return references from their parameters would like to disallow rvalues and also enforce some indication at the callsite that a reference is being captured.
+`ref readonly` parameters are ideal in such cases as they warn if used with rvalues or without any annotation at the callsite.
+
+Furthermore, there are APIs that need only read-only references but use
+- `ref` parameters since they were introduced before `in` became available and changing to `in` would be a source and binary breaking change, e.g., `QueryInterface`, or
+- `in` parameters to accept readonly references even though passing rvalues to them doesn't really make sense, e.g., `ReadOnlySpan<T>..ctor(in T value)`, or
+- `ref` parameters to disallow rvalues even though they don't mutate the passed reference, e.g., `Unsafe.IsNullRef`.
+
+These APIs could migrate to `ref readonly` parameters without breaking users.
+For details on binary compatibility, see the proposed [metadata encoding][metadata].
+Specifically, changing
+- `ref` → `ref readonly` would only be a binary breaking change for virtual methods,
+- `ref` → `in` would also be a binary breaking change for virtual methods, but not a source breaking change (because the rules change to only warn for `ref` arguments passed to `in` parameters),
+- `in` → `ref readonly` would not be a breaking change (but no callsite annotation or rvalue would result in a warning),
+  - note that this would be a source breaking change for users using older compiler versions (as they interpret `ref readonly` parameters as `ref` parameters, disallowing `in` or no annotation at the callsite) and new compiler versions with `LangVersion <= 11` (for consistency, an error will be emitted that `ref readonly` parameters are not supported).
+
+In the opposite direction, changing
+- `ref readonly` → `ref` would be potentially a source breaking change (unless only `ref` callsite annotation was used and only readonly references used as arguments), and a binary breaking change for virtual methods,
+- `ref readonly` → `in` would not be a breaking change (but `ref` callsite annotation would result in a warning).
+
+## Detailed design
+[design]: #detailed-design
+
+No changes in grammar are necessary.
+Specification will be extended to allow `ref readonly` modifiers for parameters with the same rules as specified for `in` parameters in [their proposal](https://github.com/dotnet/csharplang/blob/c8c1615fcad4ca016a97b7260ad497aad53ebc78/proposals/csharp-7.2/readonly-ref.md), except where explicitly changed in this proposal.
+
+Note that even though `ref` argument modifier is allowed for `ref readonly` parameters, nothing changes w.r.t. value kind checks, i.e.,
+- `ref` can only be used with assignable values;
```

Comment:
Does this mean I wouldn't be able to do the following? If so, it seems a bit weird:
```csharp
int a = 1;
ref readonly int b = ref a; //note we use ref for this
ref readonly int c = ref b; //and for this

//would the following 2 lines be errors?
X1(ref c);
X2(ref c); //warning here obviously

void X1(ref readonly int x) { }
void X2(in int x) { }
```

I was hoping we would be able to just use `ref` everywhere. It seems a bit odd to disallow `ref` here, since we use `ref` (not `in`) for `ref readonly` locals etc.

Reply:
Yes, `X1(ref c)` will be an error no matter whether `X1` is defined with `ref`/`in`/`ref readonly` parameter (the value kind of `c` is checked against the `ref` modifier at the callsite), you will have to use `X1(in c)` when `c` is readonly. The reason is that `ref` at the call site says "passing a mutable reference" whereas `in` says "passing a readonly reference" - `ref readonly` parameters just allow both of them.

Ref assignments are special, they use `ref` always, but it's pretty clear if it's readonly or not, since that's declared right there on the left-hand side.